File: auth0_flutter/darwin/Classes/MfaAPI/MfaExtensions.swift
import Auth0

#if os(iOS)
import Flutter
#else
import FlutterMacOS
#endif

extension FlutterError {
    convenience init(from error: MfaListAuthenticatorsError) {
        self.init(fromMfaInfo: error.info, code: error.code, statusCode: error.statusCode)
    }

    convenience init(from error: MfaEnrollmentError) {
        self.init(fromMfaInfo: error.info, code: error.code, statusCode: error.statusCode)
    }

    convenience init(from error: MfaChallengeError) {
        self.init(fromMfaInfo: error.info, code: error.code, statusCode: error.statusCode)
    }

    convenience init(from error: MFAVerifyError) {
        self.init(fromMfaInfo: error.info, code: error.code, statusCode: error.statusCode)
    }

    private convenience init(fromMfaInfo info: [String: Any], code: String, statusCode: Int) {
        let isNetworkError = (info["cause"] as? Error) != nil
        let description = info["error_description"] as? String
            ?? info["description"] as? String
            ?? code
        let details: [String: Any] = [
            "_statusCode": statusCode,
            "_errorFlags": [
                "isNetworkError": isNetworkError
            ]
        ]
        self.init(code: code, message: description, details: details)
    }
}

extension Authenticator {
    func asDictionary() -> [String: Any?] {
        return [
            "id": id,
            "type": type,
            "authenticator_type": authenticatorType,
            "active": active,
            "oob_channel": oobChannel,
            "name": name
        ]
    }
}

extension MFAChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "challenge_type": challengeType,
            "oob_code": oobCode,
            "binding_method": bindingMethod
        ]
    }
}

extension MFAEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "authenticator_type": authenticatorType,
            "oob_channel": oobChannel,
            "oob_code": oobCode,
            "binding_method": bindingMethod,
            "totp_secret": nil,
            "barcode_uri": nil,
            "recovery_codes": recoveryCodes,
            "id": nil,
            "auth_session": nil
        ]
    }
}

extension OTPMFAEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "authenticator_type": authenticatorType,
            "oob_channel": nil,
            "oob_code": nil,
            "binding_method": nil,
            "totp_secret": secret,
            "barcode_uri": barcodeUri,
            "recovery_codes": recoveryCodes,
            "id": nil,
            "auth_session": nil
        ]
    }
}

extension PushMFAEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "authenticator_type": authenticatorType,
            "oob_channel": oobChannel,
            "oob_code": oobCode,
            "binding_method": nil,
            "totp_secret": nil,
            "barcode_uri": barcodeUri,
            "recovery_codes": recoveryCodes,
            "id": nil,
            "auth_session": nil
        ]
    }
}

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/MfaExtensions.kt
package com.auth0.auth0_flutter

import com.auth0.android.authentication.mfa.MfaException
import com.auth0.android.result.Authenticator
import com.auth0.android.result.Challenge
import com.auth0.android.result.Credentials
import com.auth0.android.result.EnrollmentChallenge
import com.auth0.android.result.MfaEnrollmentChallenge
import com.auth0.android.result.OobEnrollmentChallenge
import com.auth0.android.result.RecoveryCodeEnrollmentChallenge
import com.auth0.android.result.TotpEnrollmentChallenge

fun Authenticator.toMfaMap(): Map<String, Any?> = buildMap {
    put("id", id)
    put("type", type)
    put("authenticator_type", authenticatorType)
    put("active", active)
    put("oob_channel", oobChannel)
    put("name", name)
}

fun Challenge.toMfaChallengeMap(): Map<String, Any?> = buildMap {
    put("challenge_type", challengeType)
    put("oob_code", oobCode)
    put("binding_method", bindingMethod)
}

fun EnrollmentChallenge.toMfaEnrollmentMap(): Map<String, Any?> = buildMap {
    put("id", id)
    put("auth_session", authSession)
    put("oob_code", oobCode)
    when (val challenge = this@toMfaEnrollmentMap) {
        is TotpEnrollmentChallenge -> {
            put("authenticator_type", "otp")
            put("totp_secret", challenge.manualInputCode)
            put("barcode_uri", challenge.barcodeUri)
        }
        is OobEnrollmentChallenge -> {
            put("authenticator_type", "oob")
            put("binding_method", challenge.bindingMethod)
        }
        is RecoveryCodeEnrollmentChallenge -> {
            put("authenticator_type", "recovery-code")
            put("recovery_codes", listOf(challenge.recoveryCode))
        }
        is MfaEnrollmentChallenge -> {}
        else -> {}
    }
}

fun Credentials.toMfaCredentialsMap(): Map<String, Any?> {
    val scopes = scope?.split(" ") ?: listOf()
    val formattedDate = expiresAt.toInstant().toString()
    return mapOf(
        "accessToken" to accessToken,
        "idToken" to idToken,
        "refreshToken" to refreshToken,
        "userProfile" to user.toMap(),
        "expiresAt" to formattedDate,
        "scopes" to scopes,
        "tokenType" to type
    )
}

fun MfaException.toMfaMap(): Map<String, Any> = buildMap {
    // `code` and `description` carry the actual error from the native MFA
    // SDK; surface them in the details map (in addition to the top-level
    // result.error code/message) so the Dart layer always has them.
    put("code", getCode())
    put("description", getDescription())
    put("_statusCode", statusCode)
    put(
        "_errorFlags", mapOf(
            "isNetworkError" to (getCode() == "network_error")
        )
    )
}
